### PR TITLE
Use PNG for mouse cursors

### LINKED ISSUE #740: Windows 7 + VMware SVGA II driver.  black box where cursor should be
Using vncviewer for linux we have no issues.

Using novnc we can see a black box drawn around where the cursor should be.

If we do not install the VMWare SVGA II driver, novnc works, there is no black box.

The guest is a qemu windows vm (with a vmware vga adapter), using qemu's VNC option with websocket -> tcp proxy.

Using vnc_auto.html 0.6.0.

### PR DESCRIPTION
**Pull request by CendioOssman**
The browsers' support for Microsoft's cursor format is a bit spotty,
so use the more common PNG format instead. This also allows us to
use a Canvas to generate the image, rather than coding it by hand.

Fixes #740.

### CODE REVIEW

File: core/display.js [hunk 1]
```diff
@@ -763,110 +763,52 @@
...
+        if (SUPPORTS_IMAGEDATA_CONSTRUCTOR) {
+            img = new ImageData(new Uint8ClampedArray(cur), w, h);
+        } else {
+            img = ctx.createImageData(w, h);
+            img.data.set(new Uint8ClampedArray(cur));
         }
+        ctx.clearRect(0, 0, w, h);
+        ctx.putImageData(img, 0, 0);
 
-        var url = 'data:image/x-icon;base64,' + Base64.encode(cur);
```
Comment (samhed) [inline core/display.js]: Can we remove Base64.encode now?

Comment (CendioOssman) [inline core/display.js]: Still used in `imageRect()` unfortunately.

**Comment by kanaka** [comment]
@CendioOssman This would certainly be a nice improvement. I do recall when I implemented this that browsers had mixed support for dataURLs and image formats, especially in the cursor style context. I'm sure this has improved in the years since though. Do you have any testing data? Seems like as of a year ago, IE only supported .cur and .ani image types: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_User_Interface/Using_URL_values_for_the_cursor_property In particular I would be interested if you've verified this for mobile browsers.

Apart from verifying browser compatibility, the change looks good.

**Comment by samhed** [comment]
I'm not sure if I misunderstood something here, but IE doesn't support data URI's for cursors anyway?

**Comment by kanaka** [comment]
@samhed you're right, I wasn't very clear. I was giving an example of the worst case (IE) as an example of the varied support even as of 1 year ago. But as long this does not regress any of the browsers we want to support, then I'm completely fine with it. I was just curious if we have any data points to confirm we haven't regressed (especially mobile where I've personally done basically 0 local cursor testing). When I implemented local cursor support originally (and it was a royal PITA), support was all over the place and what I ended up with was the least common denominator of functionality.

BTW, regarding the base64 issue, we could consider using window.atob/btoa for imageRect. However, this is one of those surprising cases where the JS implementation of Base64 was significantly faster than the builtin functions. And to top it off, if I'm recalling correctly (it's been a while) doing a direct comparison of simple microbenchmarks did not reveal the different as starkly. The performance difference may longer be there or it might be ever worse now with even more optimized JS JITs. But that's the reason why we have Base64 in the code base and aren't just using the builtin atob/btoa functions.

**Comment by CendioOssman** [comment]
> I was just curious if we have any data points to confirm we haven't regressed (especially mobile where I've personally done basically 0 local cursor testing).

More testing is needed. I just first wanted to confirm that it solves the issue.

> BTW, regarding the base64 issue, ...

Long term we probably want to use blobs rather than data URIs, so it might not be worth changing until then.

**Comment by DirectXMan12** [comment]
marking as v0.7.0 since this fixes a bug that we probably want fixed.

generally LGTM :+1: once testing confirms that it has broad support.

**Comment by CendioOssman** [comment]
Tested on all the major browsers and it works well.

**Pull request merged**